chasm/lib/activity/activity_tasks.go

```diff
@@ -0,0 +1,82 @@
+package activity
+
+import (
+	"context"
+
+	"go.temporal.io/server/api/matchingservice/v1"
+	"go.temporal.io/server/chasm"
+	"go.temporal.io/server/chasm/lib/activity/gen/activitypb/v1"
+	"go.temporal.io/server/common/resource"
+	"go.uber.org/fx"
+)
+
+type activityDispatchTaskExecutorOptions struct {
+	fx.In
+
+	MatchingClient resource.MatchingClient
+}
+
+type activityDispatchTaskExecutor struct {
+	opts activityDispatchTaskExecutorOptions
+}
+
+func newActivityDispatchTaskExecutor(opts activityDispatchTaskExecutorOptions) *activityDispatchTaskExecutor {
+	return &activityDispatchTaskExecutor{
+		opts,
+	}
+}
+
+func (e *activityDispatchTaskExecutor) Validate(
+	ctx chasm.Context,
+	activity *Activity,
+	_ chasm.TaskAttributes,
+	task *activitypb.ActivityDispatchTask,
+) (bool, error) {
+	attempt, err := activity.Attempt.Get(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// TODO make sure we handle resets when we support them, as they will reset the attempt count
+	if activity.State() != activitypb.ACTIVITY_EXECUTION_STATUS_SCHEDULED || task.Attempt != attempt.Count {
```

Comment:
I typically like to model this using the state machine transitions, basically checking if the machine can transition to Started (using the `Possible()` method) but this is also fine.